modules/generator/generator_kafka.go

```diff
@@ -0,0 +1,231 @@
+				// Starting up or shutting down
+				continue
+			}
+			err := g.readKafka(ctx)
+			if err != nil {
+				level.Error(g.logger).Log("msg", "readKafka failed", "err", err)
+				continue
+			}
+		case <-g.kafkaStop:
+			return
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (g *Generator) readKafka(ctx context.Context) error {
+	fallback := time.Now().Add(-time.Minute)
+
+	groupLag, err := getGroupLag(
+		ctx,
+		kadm.NewClient(g.kafkaClient),
+		g.cfg.Ingest.Kafka.Topic,
+		g.cfg.Ingest.Kafka.ConsumerGroup,
+		fallback,
+	)
+	if err != nil {
+		return fmt.Errorf("failed to get group lag: %w", err)
+	}
+
+	assignedPartitions := g.getAssignedActivePartitions()
+
+	for _, partition := range assignedPartitions {
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
+
+		partitionLag, ok := groupLag.Lookup(g.cfg.Ingest.Kafka.Topic, partition)
+		if !ok {
+			return fmt.Errorf("lag for partition %d not found", partition)
+		}
+
+		if partitionLag.Lag <= 0 {
+			// Nothing to consume
+			continue
+		}
+
+		err := g.consumePartition(ctx, partition, partitionLag)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (g *Generator) consumePartition(ctx context.Context, partition int32, lag kadm.GroupMemberLag) error {
+	d := ingest.NewDecoder()
+
+	// We always rewind the partition's offset to the commit offset by reassigning the partition to the client (this triggers partition assignment).
+	// This is so the cycle started exactly at the commit offset, and not at what was (potentially over-) consumed previously.
+	// In the end, we remove the partition from the client (refer to the defer below) to guarantee the client always consumes
+	// from one partition at a time. I.e. when this partition is consumed, we start consuming the next one.
+	g.kafkaClient.AddConsumePartitions(map[string]map[int32]kgo.Offset{
+		g.cfg.Ingest.Kafka.Topic: {
+			partition: kgo.NewOffset().At(lag.Commit.At),
+		},
+	})
+	defer g.kafkaClient.RemoveConsumePartitions(map[string][]int32{g.cfg.Ingest.Kafka.Topic: {partition}})
+
+	fetches := g.kafkaClient.PollFetches(ctx)
+	fetches.EachError(func(_ string, _ int32, err error) {
+		if !errors.Is(err, context.Canceled) {
+			level.Error(g.logger).Log("msg", "failed to fetch records", "err", err)
+		}
+	})
+
+	for _, f := range fetches {
+		for _, t := range f.Topics {
+			for _, p := range t.Partitions {
+				if p.Partition != partition {
+					continue
+				}
+
+				for _, r := range p.Records {
+					tenant := string(r.Key)
+
+					i, err := g.getOrCreateInstance(tenant)
+					if err != nil {
+						return err
+					}
+
+					d.Reset()
+					req, err := d.Decode(r.Value)
+					if err != nil {
+						return err
+					}
+
+					for _, tr := range req.Traces {
+						trace := &tempopb.Trace{}
+						err = trace.Unmarshal(tr.Slice)
+						if err != nil {
+							return err
+						}
+
+						i.pushSpansFromQueue(ctx, &tempopb.PushSpansRequest{
+							Batches: trace.ResourceSpans,
+						})
+					}
+				}
+			}
+		}
+	}
```

Comment:
These loops can be simplified since we're direct consuming and all the fetches belong to the topic/partition defined above

```go
for recIter := fetches.RecordIter(); !recIter.Done(); {
  <code>
}
```